src/internal/util/logger.ts

```diff
@@ -1,119 +1,286 @@
 import * as cluster from "cluster";
 import * as fs from "fs-extra";
-import * as stringify from "json-stringify-safe";
 import * as _ from "lodash";
+import * as logform from "logform";
 import * as os from "os";
 import * as p from "path";
 import * as serializeError from "serialize-error";
-import * as stripAnsi from "strip-ansi";
 import * as winston from "winston";
+import { Configuration } from "../../configuration";
 import * as context from "./cls";
 
-export let LoggingConfig = {
-    format: "logger",
+const winstonLogger = winston.createLogger({
+    level: "debug",
+    exitOnError: false,
+    silent: true,
+});
+
+/**
+ * Global logger instance
+ * By default all logging to this logger will be suppressed.
+ * Call configureLogging to set up console and file transports.
+ */
+export const logger: Logger = winstonLogger;
+
+// Save console log methods
+const GlobalConsoleMethods = {
+    error: console.error,
+    warn: console.warn,
+    info: console.info,
+    debug: console.debug,
+    log: console.log,
+    trace: console.trace,
 };
 
-export function setLogLevel(level: string) {
-    winstonLogger.transports.console.level = level;
+/**
+ * Constants for the logging format
+ */
+export enum LoggingFormat {
+    /**
+     * Print only the message provided to the logger methods
+     */
+    None,
+    /**
+     * Print level and message
+     */
+    Simple,
+    /**
+     * Print the full client related metadata
+     */
+    Full,
 }
 
-export function addFileTransport(filename: string, level: string) {
-    const path = p.resolve(filename);
-    fs.mkdirsSync(p.dirname(path));
+/**
+ * Configuration for the logging sub-system to be passed to configureLogging
+ */
+export interface LoggingConfiguration {
+    console: {
+        /**
+         * Enable logging to console
+         */
+        enabled: boolean,
+        /**
+         * Set log level for the console
+         */
+        level?: string,
+        /**
+         * Set format of console log
+         */
+        format?: LoggingFormat,
+        /**
+         * Redirect console.log methods to logger sub-system
+         */
+        redirect?: boolean,
+    };
+    file: {
+        /**
+         * Enable file logging
+         */
+        enabled: boolean,
+        /**
+         * Set log level for file logging
+         */
+        level?: string,
+        /**
+         * Set file name and path to log into
+         */
+        filename?: string,
+        /**
+         * Set format for file logging
+         */
+        format?: LoggingFormat,
+    };
+}
+
+/**
+ * Logging configuration to simply pass through log message to the console
+ */
+export const NoLoggingConfiguration: LoggingConfiguration = {
```

Comment:
`NoLoggingConfiguration` seems like a misnomer to me. When I first saw it above I thought it meant all log messages would be eaten, i.e., no log statements would be output. Perhaps `PlainLoggingConfiguration` would be better?

Reply:
Right, I'll change it.